### PR TITLE
Fixed #4295

### LINKED ISSUE #4295: ES Modules broken on npm@7
# Environment

Knex version: 0.21.17
Database + version: N/A
OS: N/A

# Bug

1. Explain what kind of behaviour you are getting and how you think it should do

When setting `type: 'module'` in `package.json` knex fails to use ESM to load the migration files on `npm >= 7.0.0`. This works on lower versions. The reason being that npm has removed the environment variable which knex uses to determine if it should use ESM or CommonJS (https://github.com/knex/knex/blob/b20a31fe5d0256234e1542c1ec4cbe2082f9970e/lib/migrations/util/import-file.js#L2). See this NPM RFC for more details https://github.com/npm/rfcs/blob/a044b355af60310b05ad1995fdc1cf672243a4e0/implemented/0021-reduce-lifecycle-script-environment.md. If the file is named `knexfile.mjs` it works as expected since the filename check triggers the usage of ESM.

2. Error message

```
Error [ERR_REQUIRE_ESM]: Must use import to load ES Module: /project-root/config/knexfile.js
require() of ES modules is not supported.
require() of /project-root/node_modules/knex/lib/util/import-file.js is an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which defines all .js files in that package scope as ES modules.
Instead rename knexfile.js to end in .cjs, change the requiring code to use import(), or remove "type": "module" from /project-root/package.json.

    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1080:13)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Module.require (internal/modules/cjs/loader.js:952:19)
    at require (internal/modules/cjs/helpers.js:88:18)
    at importFile (/project-root/node_modules/knex/lib/util/import-file.js:13:7)
    at openKnexfile (/project-root/node_modules/knex/bin/cli.js:26:22)
    at initKnex (/project-root/node_modules/knex/bin/cli.js:47:13)
    at Command.<anonymous> (/project-root/node_modules/knex/bin/cli.js:170:32)
    at Command.listener [as _actionHandler] (/project-root/node_modules/commander/index.js:426:31)
```

3. Reduced test code, for example in https://npm.runkit.com/knex or if it needs real
   database connection to MySQL or PostgreSQL, then single file example which initializes
   needed data and demonstrates the problem.
   
  All that's needed is a project with `type: 'module'` in the `package.json` and a knexfile ending in `.js`.
 

npm's recommended migration route is to use the variable `npm_package_json` which points to the `package.json`, load the JSON and then read values from there.

### PR DESCRIPTION
**Pull request by richardsimko**

### CODE REVIEW

**Review by kibertoad — APPROVED** [review]

**Comment by kibertoad** [comment]
Thank you!

File: lib/migrate/Migrator.js [hunk 1]
```diff
@@ -80,12 +80,16 @@ class Migrator {
...
+      !(
+        await Promise.all(
+          migrations.map(async (migration) => {
+            const migrationContents = await this.config.migrationSource.getMigration(
+              migration
+            );
+            return !this._useTransaction(migrationContents);
+          })
+        )
+      ).some((isTransactionUsed) => isTransactionUsed);
```
Comment (falkenhawk) [inline lib/migrate/Migrator.js:92]: were the changes in `Migrator` intended in this PR?

Comment (falkenhawk) [reply lib/migrate/Migrator.js:92]: Okay I can see more details in https://github.com/knex/knex/pull/4296

Comment (kibertoad) [reply lib/migrate/Migrator.js:92]: @falkenhawk Did that cause some breaking changes for you?

Comment (falkenhawk) [reply lib/migrate/Migrator.js:92]: I can't tell yet, I noticed this while I was looking through the changes after I got notified by renovatebot about the new knex version 0.21.18... and not having this explicitly listed in changelog confused me and stopped me from upgrading, as it seemed to me to be something extra than just a fix for #4296 - and I could not afford breaking things at that moment 😅

If I decide to upgrade in the following days and will run into some issues, I will let you know then.

Comment (kibertoad) [reply lib/migrate/Migrator.js:92]: It should behave same way for the end user in most cases, and also handle ESM case better, see https://github.com/knex/knex/pull/4296#issuecomment-779330504 for the reason of more methods becoming async. If you are not using ESM, this upgrade won't do anything useful for you.

**Pull request merged**